File: lectures/lecture-04-why-one-giant-instruction-file-fails/exercises/exercise-02-snr-calculator/solution/python/main.py
"""snr-calculator exercise, Python solution.

Computes per-task instruction signal-to-noise for one tree: relevant
instruction lines (tag-matched) over loaded lines. Contract: ../../SPEC.md.
"""

from __future__ import annotations

import json
import re
import sys
from pathlib import Path

RULE_RE = re.compile(r"^- \[([a-z]+)(!?)\] (.+)$")


def read_lines(path: Path) -> list[str]:
    lines = path.read_text(encoding="utf-8").split("\n")
    if lines and lines[-1] == "":
        lines = lines[:-1]
    return lines


def relevant_count(lines: list[str], topics: list[str]) -> int:
    """An instruction line is relevant when its TAG names one of the task's
    topics. Prose that mentions a topic word is context cost, not signal."""
    count = 0
    for line in lines:
        match = RULE_RE.match(line)
        if match and match.group(1) in topics:
            count += 1
    return count


def snr_report(tree: Path, tasks: list[dict]) -> dict:
    lines = read_lines(tree / "AGENTS.md")
    loaded = len(lines)
    rows = []
    snr_total = 0.0
    for task in tasks:
        relevant = relevant_count(lines, task["topics"])
        snr = relevant / loaded if loaded else 0.0
        snr_total += snr
        rows.append({
            "id": task["id"],
            "loaded_lines": loaded,
            "relevant_lines": relevant,
            "snr": snr,
        })
    return {
        "tasks": rows,
        "mean_snr": snr_total / len(rows) if rows else 0.0,
    }


def main(argv: list[str]) -> int:
    if len(argv) != 3:
        print("usage: main.py <tree-dir> <tasks.json>", file=sys.stderr)
        return 2
    tree = Path(argv[1])
    if not (tree / "AGENTS.md").is_file():
        print(f"error: no AGENTS.md in {tree}", file=sys.stderr)
        return 2
    try:
        tasks = json.loads(Path(argv[2]).read_text(encoding="utf-8"))["tasks"]
    except OSError as error:
        print(f"error: cannot read tasks: {error}", file=sys.stderr)
        return 2
    print(json.dumps(snr_report(tree, tasks), indent=2))
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: lectures/lecture-04-why-one-giant-instruction-file-fails/exercises/exercise-02-snr-calculator/solution/typescript/main.ts
// snr-calculator exercise, TypeScript solution.
//
// Computes per-task instruction signal-to-noise for one tree: relevant
// instruction lines (tag-matched) over loaded lines. Contract: ../../SPEC.md.

import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { pathToFileURL } from "node:url";

const RULE_RE = /^- \[([a-z]+)(!?)\] (.+)$/;

interface Task {
  readonly id: string;
  readonly topics: readonly string[];
}

function readLines(path: string): string[] {
  let lines = readFileSync(path, "utf8").split(/\r?\n/);
  if (lines.length > 0 && lines[lines.length - 1] === "") lines = lines.slice(0, -1);
  return lines;
}

export function relevantCount(lines: readonly string[], topics: readonly string[]): number {
  // An instruction line is relevant when its TAG names one of the task's
  // topics. Prose that mentions a topic word is context cost, not signal.
  let count = 0;
  for (const line of lines) {
    const match = line.match(RULE_RE);
    if (match && match[1] && topics.includes(match[1])) count += 1;
  }
  return count;
}

export function snrReport(tree: string, tasks: readonly Task[]): Record<string, unknown> {
  const lines = readLines(join(tree, "AGENTS.md"));
  const loaded = lines.length;
  const rows = [];
  let snrTotal = 0;
  for (const task of tasks) {
    const relevant = relevantCount(lines, task.topics);
    const snr = loaded > 0 ? relevant / loaded : 0;
    snrTotal += snr;
    rows.push({ id: task.id, loaded_lines: loaded, relevant_lines: relevant, snr });
  }
  return {
    tasks: rows,
    mean_snr: rows.length > 0 ? snrTotal / rows.length : 0,
  };
}

function main(argv: readonly string[]): number {
  const tree = argv[2];
  const tasksPath = argv[3];
  if (!tree || !tasksPath || argv.length !== 4) {
    console.error("usage: main.ts <tree-dir> <tasks.json>");
    return 2;
  }
  if (!existsSync(join(tree, "AGENTS.md"))) {
    console.error(`error: no AGENTS.md in ${tree}`);
    return 2;
  }
  let tasks: Task[];
  try {
    tasks = (JSON.parse(readFileSync(tasksPath, "utf8")) as { tasks: Task[] }).tasks;
  } catch (error) {
    console.error(`error: cannot read tasks: ${String(error)}`);
    return 2;
  }
  process.stdout.write(JSON.stringify(snrReport(tree, tasks), null, 2) + "\n");
  return 0;
}

if (import.meta.url === pathToFileURL(process.argv[1] ?? "").href) {
  process.exit(main(process.argv));
}
